File: packages/runtime-ts/tests/compliance.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { BitWriter } from '../src/bit-writer.js';
import { BitReader } from '../src/bit-reader.js';

const __dirname = dirname(fileURLToPath(import.meta.url));
const vectorsDir = join(__dirname, '..', '..', '..', 'compliance', 'vectors');

/** Convert hex string to Uint8Array. */
function hexToBytes(hex: string): Uint8Array {
  const bytes = new Uint8Array(hex.length / 2);
  for (let i = 0; i < hex.length; i += 2) {
    bytes[i / 2] = parseInt(hex.substring(i, i + 2), 16);
  }
  return bytes;
}

/** Convert Uint8Array to hex string. */
function toHex(bytes: Uint8Array): string {
  return Array.from(bytes)
    .map((b) => b.toString(16).padStart(2, '0'))
    .join('');
}

interface Vector {
  name: string;
  schema: string;
  type: string;
  value: Record<string, unknown>;
  expected_bytes: string;
  notes?: string;
}

/**
 * Encode a single field value using a BitWriter based on the schema type info.
 * This parses the schema string minimally to determine field types and order.
 */
function encodeValue(
  schema: string,
  value: Record<string, unknown>,
): Uint8Array {
  const fields = parseFields(schema);
  const w = new BitWriter();

  for (const field of fields) {
    const v = value[field.name];
    writeField(w, field.type, v);
  }

  return w.finish();
}

/**
 * Decode fields from bytes using a BitReader based on the schema.
 */
function decodeValue(
  schema: string,
  bytes: Uint8Array,
): Record<string, unknown> {
  const fields = parseFields(schema);
  const r = new BitReader(bytes);
  const result: Record<string, unknown> = {};

  for (const field of fields) {
    result[field.name] = readField(r, field.type);
  }

  return result;
}

interface FieldDef {
  name: string;
  type: string;
}

/**
 * Minimal schema parser: extracts field names and types from a message definition.
 * Handles schemas like: "namespace test.prim\nmessage M { v @0 : bool }"
 */
function parseFields(schema: string): FieldDef[] {
  // Find message body between { }
  const braceStart = schema.indexOf('{');
  const braceEnd = schema.lastIndexOf('}');
  if (braceStart === -1 || braceEnd === -1) return [];

  const body = schema.substring(braceStart + 1, braceEnd).trim();
  if (body.length === 0) return [];

  const fields: FieldDef[] = [];
  // Match: name @N : type
  const fieldRegex = /(\w+)\s+@\d+\s*:\s*(\w+)/g;
  let match;
  while ((match = fieldRegex.exec(body)) !== null) {
    fields.push({ name: match[1], type: match[2] });
  }

  return fields;
}

function writeField(
  w: BitWriter,
  type: string,
  value: unknown,
): void {
  // Handle sub-byte types like u1, u3, u5, u6
  const subByteMatch = type.match(/^u(\d+)$/);
  if (subByteMatch) {
    const bits = parseInt(subByteMatch[1], 10);
    if (bits < 8 && bits >= 1) {
      w.writeBits(value as number, bits);
      return;
    }
  }

  switch (type) {
    case 'bool':
      w.writeBool(value as boolean);
      break;
    case 'u8':
      w.writeU8(value as number);
      break;
    case 'u16':
      w.writeU16(value as number);
      break;
    case 'u32':
      w.writeU32(value as number);
      break;
    case 'u64':
      w.writeU64(BigInt(value as number));
      break;
    case 'i8':
      w.writeI8(value as number);
      break;
    case 'i16':
      w.writeI16(value as number);
      break;
    case 'i32':
      w.writeI32(value as number);
      break;
    case 'i64':
      w.writeI64(BigInt(value as number));
      break;
    case 'f32': {
      const v = value as number | string;
      if (v === 'NaN' || (typeof v === 'string' && v === 'NaN')) {
        w.writeF32(NaN);
      } else if (typeof v === 'string') {
        w.writeF32(parseFloat(v));
      } else {
        w.writeF32(Math.fround(v));
      }
      break;
    }
    case 'f64': {
      const v = value as number | string;
      if (v === 'NaN' || (typeof v === 'string' && v === 'NaN')) {
        w.writeF64(NaN);
      } else if (typeof v === 'string' && v === '-0.0') {
        w.writeF64(-0.0);
      } else if (typeof v === 'string') {
        w.writeF64(parseFloat(v));
      } else {
        w.writeF64(v);
      }
      break;
    }
    case 'string':
      w.writeString(value as string);
      break;
    default:
      throw new Error(`Unsupported type: ${type}`);
  }
}

function readField(r: BitReader, type: string): unknown {
  // Handle sub-byte types like u1, u3, u5, u6
  const subByteMatch = type.match(/^u(\d+)$/);
  if (subByteMatch) {
    const bits = parseInt(subByteMatch[1], 10);
    if (bits < 8 && bits >= 1) {
      return r.readBits(bits);
    }
  }

  switch (type) {
    case 'bool':
      return r.readBool();
    case 'u8':
      return r.readU8();
    case 'u16':
      return r.readU16();
    case 'u32':
      return r.readU32();
    case 'u64':
      return r.readU64();
    case 'i8':
      return r.readI8();
    case 'i16':
      return r.readI16();
    case 'i32':
      return r.readI32();
    case 'i64':
      return r.readI64();
    case 'f32':
      return r.readF32();
    case 'f64':
      return r.readF64();
    case 'string':
      return r.readString();
    default:
      throw new Error(`Unsupported type: ${type}`);
  }
}

/**
 * Compare decoded value against expected, handling special cases
 * like NaN, -0.0, and bigint.
 */
function valuesMatch(
  actual: unknown,
  expected: unknown,
  type: string,
): boolean {
  if (type === 'f32' || type === 'f64') {
    if (expected === 'NaN') {
      return Number.isNaN(actual as number);
    }
    if (expected === '-0.0') {
      return Object.is(actual, -0);
    }
    if (typeof expected === 'string') {
      return (actual as number) === parseFloat(expected);
    }
  }
  if (type === 'u64' || type === 'i64') {
    return (actual as bigint) === BigInt(expected as number);
  }
  return actual === expected;
}

describe('Compliance: primitives.json', () => {
  const vectors: Vector[] = JSON.parse(
    readFileSync(join(vectorsDir, 'primitives.json'), 'utf-8'),
  );

  for (const vec of vectors) {
    describe(vec.name, () => {
      it('encode matches expected bytes', () => {
        const encoded = encodeValue(vec.schema, vec.value);
        expect(toHex(encoded)).toBe(vec.expected_bytes);
      });

      it('decode matches expected value', () => {
        const bytes = hexToBytes(vec.expected_bytes);
        const decoded = decodeValue(vec.schema, bytes);
        const fields = parseFields(vec.schema);
        for (const field of fields) {
          expect(
            valuesMatch(decoded[field.name], vec.value[field.name], field.type),
          ).toBe(true);
        }
      });
    });
  }
});

describe('Compliance: sub_byte.json', () => {
  const vectors: Vector[] = JSON.parse(
    readFileSync(join(vectorsDir, 'sub_byte.json'), 'utf-8'),
  );

  for (const vec of vectors) {
    describe(vec.name, () => {
      it('encode matches expected bytes', () => {
        const encoded = encodeValue(vec.schema, vec.value);
        expect(toHex(encoded)).toBe(vec.expected_bytes);
      });

      it('decode matches expected value', () => {
        const bytes = hexToBytes(vec.expected_bytes);
        const decoded = decodeValue(vec.schema, bytes);
        const fields = parseFields(vec.schema);
        for (const field of fields) {
          expect(
            valuesMatch(decoded[field.name], vec.value[field.name], field.type),
          ).toBe(true);
        }
      });
    });
  }
});

describe('Compliance: messages.json', () => {
  const vectors: Vector[] = JSON.parse(
    readFileSync(join(vectorsDir, 'messages.json'), 'utf-8'),
  );

  for (const vec of vectors) {
    describe(vec.name, () => {
      it('encode matches expected bytes', () => {
        const encoded = encodeValue(vec.schema, vec.value);
        expect(toHex(encoded)).toBe(vec.expected_bytes);
      });

      // For messages with fields, also test decode
      if (Object.keys(vec.value).length > 0) {
        it('decode matches expected value', () => {
          const bytes = hexToBytes(vec.expected_bytes);
          const decoded = decodeValue(vec.schema, bytes);
          const fields = parseFields(vec.schema);
          for (const field of fields) {
            expect(
              valuesMatch(
                decoded[field.name],
                vec.value[field.name],
                field.type,
              ),
            ).toBe(true);
          }
        });
      }
    });
  }
});

describe('Compliance: optionals.json', () => {
  interface OptionalVector {
    name: string;
    schema: string;
    type: string;
    value: Record<string, unknown>;
    expected_bytes: string;
    notes?: string;
  }

  const vectors: OptionalVector[] = JSON.parse(
    readFileSync(join(vectorsDir, 'optionals.json'), 'utf-8'),
  );

  for (const vec of vectors) {
    describe(vec.name, () => {
      it('encode matches expected bytes', () => {
        const w = new BitWriter();
        const v = vec.value.v;
        if (v === null) {
          w.writeBool(false);
        } else {
          w.writeBool(true);
          w.flushToByteBoundary();
          w.writeU32(v as number);
        }
        expect(toHex(w.finish())).toBe(vec.expected_bytes);
      });

      it('decode matches expected value', () => {
        const r = new BitReader(hexToBytes(vec.expected_bytes));
        const present = r.readBool();
        if (!present) {
          expect(vec.value.v).toBeNull();
        } else {
          r.flushToByteBoundary();
          const val = r.readU32();
          expect(val).toBe(vec.value.v);
        }
      });
    });
  }
});

describe('Compliance: enums.json', () => {
  interface EnumVector {
    name: string;
    schema: string;
    type: string;
    value: Record<string, unknown>;
    expected_bytes: string;
    notes?: string;
  }

  const vectors: EnumVector[] = JSON.parse(
    readFileSync(join(vectorsDir, 'enums.json'), 'utf-8'),
  );

  const variantMap: Record<string, number> = { Active: 0, Inactive: 1 };
  const indexToVariant = ['Active', 'Inactive'];

  for (const vec of vectors) {
    describe(vec.name, () => {
      it('encode matches expected bytes', () => {
        const w = new BitWriter();
        const variant = vec.value.v as string;
        w.writeBits(variantMap[variant], 1);
        expect(toHex(w.finish())).toBe(vec.expected_bytes);
      });

      it('decode matches expected value', () => {
        const r = new BitReader(hexToBytes(vec.expected_bytes));
        const idx = r.readBits(1);
        expect(indexToVariant[idx]).toBe(vec.value.v);
      });
    });
  }
});

describe('Compliance: unions.json', () => {
  interface UnionVector {
    name: string;
    schema: string;
    type: string;
    value: Record<string, unknown>;
    expected_bytes: string;
    notes?: string;
  }

  const vectors: UnionVector[] = JSON.parse(
    readFileSync(join(vectorsDir, 'unions.json'), 'utf-8'),
  );

  for (const vec of vectors) {
    describe(vec.name, () => {
      it('encode matches expected bytes', () => {
        const w = new BitWriter();
        const unionVal = vec.value.v as Record<string, unknown>;
        const variant = unionVal.variant as string;
        const discriminant = variant === 'Circle' ? 0 : 1;

        const pw = new BitWriter();
        if (variant === 'Circle') {
          pw.writeF32(Math.fround(unionVal.radius as number));
        } else {
          pw.writeF32(Math.fround(unionVal.w as number));
          pw.writeF32(Math.fround(unionVal.h as number));
        }
        const payload = pw.finish();

        w.writeLeb128(discriminant);
        w.writeLeb128(payload.length);
        w.writeRawBytes(payload);

        expect(toHex(w.finish())).toBe(vec.expected_bytes);
      });

      it('decode matches expected value', () => {
        const r = new BitReader(hexToBytes(vec.expected_bytes));
        const discriminant = r.readLeb128();
        const payloadLen = r.readLeb128();
        const payloadBytes = r.readRawBytes(payloadLen);
        const pr = new BitReader(payloadBytes);

        const unionVal = vec.value.v as Record<string, unknown>;
        if (discriminant === 0) {
          const radius = pr.readF32();
          expect(radius).toBeCloseTo(unionVal.radius as number, 5);
        } else {
          const w = pr.readF32();
          const h = pr.readF32();
          expect(w).toBeCloseTo(unionVal.w as number, 5);
          expect(h).toBeCloseTo(unionVal.h as number, 5);
        }
      });
    });
  }
});

describe('Compliance: arrays_maps.json', () => {
  interface ArrayMapVector {
    name: string;
    schema: string;
    type: string;
    value: Record<string, unknown>;
    expected_bytes: string;
    notes?: string;
  }

  const vectors: ArrayMapVector[] = JSON.parse(
    readFileSync(join(vectorsDir, 'arrays_maps.json'), 'utf-8'),
  );

  for (const vec of vectors) {
    describe(vec.name, () => {
      it('encode matches expected bytes', () => {
        const w = new BitWriter();
        const v = vec.value.v;

        if (Array.isArray(v)) {
          w.writeLeb128(v.length);
          for (const elem of v) {
            w.writeU32(elem as number);
          }
        } else if (typeof v === 'object' && v !== null) {
          const entries = Object.entries(v as Record<string, unknown>);
          w.writeLeb128(entries.length);
          for (const [key, val] of entries) {
            w.writeString(key);
            w.writeU32(val as number);
          }
        }

        expect(toHex(w.finish())).toBe(vec.expected_bytes);
      });

      it('decode matches expected value', () => {
        const r = new BitReader(hexToBytes(vec.expected_bytes));
        const v = vec.value.v;

        if (Array.isArray(v)) {
          const count = r.readLeb128();
          expect(count).toBe(v.length);
          for (let i = 0; i < count; i++) {
            expect(r.readU32()).toBe(v[i]);
          }
        } else if (typeof v === 'object' && v !== null) {
          const expected = v as Record<string, unknown>;
          const count = r.readLeb128();
          expect(count).toBe(Object.keys(expected).length);
          for (let i = 0; i < count; i++) {
            const key = r.readString();
            const val = r.readU32();
            expect(expected[key]).toBe(val);
          }
        }
      });
    });
  }
});

describe('Compliance: evolution.json', () => {
  interface EvolutionVector {
    name: string;
    schema_v1: string;
    schema_v2: string;
    type: string;
    value_v1?: Record<string, unknown>;
    value_v2?: Record<string, unknown>;
    encoded_v1?: string;
    encoded_v2?: string;
    decoded_as_v1?: Record<string, unknown>;
    decoded_as_v2?: Record<string, unknown>;
    notes?: string;
  }

  const vectors: EvolutionVector[] = JSON.parse(
    readFileSync(join(vectorsDir, 'evolution.json'), 'utf-8'),
  );

  it('v1 encode produces expected bytes', () => {
    const vec = vectors.find(
      (v) => v.name === 'v1_encode_v2_decode_appended_field',
    )!;
    const w = new BitWriter();
    w.writeU32(vec.value_v1!.x as number);
    expect(toHex(w.finish())).toBe(vec.encoded_v1);
  });

  it('v1 bytes decoded as v2 fills default for missing field', () => {
    const vec = vectors.find(
      (v) => v.name === 'v1_encode_v2_decode_appended_field',
    )!;
    const r = new BitReader(hexToBytes(vec.encoded_v1!));
    const x = r.readU32();
    const y = r.remaining() >= 2 ? r.readU16() : 0;
    expect(x).toBe(vec.decoded_as_v2!.x);
    expect(y).toBe(vec.decoded_as_v2!.y);
  });

  it('v2 encode produces expected bytes', () => {
    const vec = vectors.find(
      (v) => v.name === 'v2_encode_v1_decode_trailing_ignored',
    )!;
    const w = new BitWriter();
    w.writeU32(vec.value_v2!.x as number);
    w.writeU16(vec.value_v2!.y as number);
    expect(toHex(w.finish())).toBe(vec.encoded_v2);
  });

  it('v2 bytes decoded as v1 ignores trailing bytes', () => {
    const vec = vectors.find(
      (v) => v.name === 'v2_encode_v1_decode_trailing_ignored',
    )!;
    const r = new BitReader(hexToBytes(vec.encoded_v2!));
    const x = r.readU32();
    expect(x).toBe(vec.decoded_as_v1!.x);
    expect(r.remaining()).toBeGreaterThan(0);
  });
});

File: packages/runtime-go/compliance_test.go
package vexil

import (
	"encoding/hex"
	"encoding/json"
	"math"
	"os"
	"strings"
	"testing"
)

// primitiveVector represents a compliance test vector for primitive types.
type primitiveVector struct {
	Name          string                 `json:"name"`
	Schema        string                 `json:"schema"`
	Type          string                 `json:"type"`
	Value         map[string]interface{} `json:"value"`
	ExpectedBytes string                 `json:"expected_bytes"`
	Notes         string                 `json:"notes"`
}

// subByteVector is the same structure as primitiveVector.
type subByteVector = primitiveVector

// messageVector is the same structure as primitiveVector.
type messageVector = primitiveVector

func loadVectors[T any](t *testing.T, filename string) []T {
	t.Helper()
	data, err := os.ReadFile("../../compliance/vectors/" + filename)
	if err != nil {
		t.Fatalf("failed to read %s: %v", filename, err)
	}
	var vectors []T
	if err := json.Unmarshal(data, &vectors); err != nil {
		t.Fatalf("failed to parse %s: %v", filename, err)
	}
	return vectors
}

func hexToBytes(t *testing.T, h string) []byte {
	t.Helper()
	b, err := hex.DecodeString(h)
	if err != nil {
		t.Fatalf("invalid hex %q: %v", h, err)
	}
	return b
}

// encodeValue writes a single value to the BitWriter based on the field schema type.
func encodeField(t *testing.T, w *BitWriter, schema string, fieldName string, value interface{}) {
	t.Helper()
	fieldType := extractFieldType(schema, fieldName)
	switch fieldType {
	case "bool":
		w.WriteBool(value.(bool))
	case "u8":
		w.WriteU8(uint8(toFloat64(value)))
	case "u16":
		w.WriteU16(uint16(toFloat64(value)))
	case "u32":
		w.WriteU32(uint32(toFloat64(value)))
	case "u64":
		w.WriteU64(uint64(toFloat64(value)))
	case "i8":
		w.WriteI8(int8(toFloat64(value)))
	case "i16":
		w.WriteI16(int16(toFloat64(value)))
	case "i32":
		w.WriteI32(int32(toFloat64(value)))
	case "i64":
		w.WriteI64(int64(toFloat64(value)))
	case "f32":
		s, ok := value.(string)
		if ok && s == "NaN" {
			w.WriteF32(float32(math.NaN()))
		} else if ok && s == "-0.0" {
			w.WriteF32(math.Float32frombits(0x80000000))
		} else {
			w.WriteF32(float32(toFloat64(value)))
		}
	case "f64":
		s, ok := value.(string)
		if ok && s == "NaN" {
			w.WriteF64(math.NaN())
		} else if ok && s == "-0.0" {
			w.WriteF64(math.Float64frombits(0x8000000000000000))
		} else {
			w.WriteF64(toFloat64(value))
		}
	case "string":
		w.WriteString(value.(string))
	default:
		// Handle sub-byte types like u1, u3, u5, u6
		if len(fieldType) >= 2 && fieldType[0] == 'u' {
			bits := uint8(0)
			for _, c := range fieldType[1:] {
				bits = bits*10 + uint8(c-'0')
			}
			w.WriteBits(uint64(toFloat64(value)), bits)
		} else {
			t.Fatalf("unsupported field type %q", fieldType)
		}
	}
}

func toFloat64(v interface{}) float64 {
	switch val := v.(type) {
	case float64:
		return val
	case int:
		return float64(val)
	case int64:
		return float64(val)
	default:
		return 0
	}
}

// extractFieldType extracts the type of a field from a simple schema string.
// Handles schemas like "message M { v @0 : bool  x @1 : u32 }"
func extractFieldType(schema string, fieldName string) string {
	// Tokenize the body content
	braceStart := strings.Index(schema, "{")
	braceEnd := strings.LastIndex(schema, "}")
	if braceStart == -1 || braceEnd == -1 {
		return ""
	}
	body := schema[braceStart+1 : braceEnd]
	tokens := strings.Fields(body)
	// Find: fieldName @N : type
	for i := 0; i < len(tokens)-3; i++ {
		if tokens[i] == fieldName && len(tokens[i+1]) >= 2 && tokens[i+1][0] == '@' && tokens[i+2] == ":" {
			return tokens[i+3]
		}
	}
	return ""
}

// extractFieldNames extracts field names in order from a schema message definition.
// Handles both multi-line and single-line schemas like "message M { v @0 : bool }"
func extractFieldNames(schema string) []string {
	var names []string
	// Use regex-like approach: find all "name @N : type" patterns
	// First, extract content between { and }
	braceStart := strings.Index(schema, "{")
	braceEnd := strings.LastIndex(schema, "}")
	if braceStart == -1 || braceEnd == -1 || braceEnd <= braceStart {
		return names
	}
	body := schema[braceStart+1 : braceEnd]
	// Split by whitespace and find field definitions
	// Fields look like: name @N : type
	// We split on multiple spaces to handle inline definitions
	tokens := strings.Fields(body)
	for i := 0; i < len(tokens); i++ {
		// Look for @N pattern (starts with @, followed by digits)
		if len(tokens[i]) >= 2 && tokens[i][0] == '@' {
			isOrdinal := true
			for _, c := range tokens[i][1:] {
				if c < '0' || c > '9' {
					isOrdinal = false
					break
				}
			}
			if isOrdinal && i > 0 {
				// Previous token is the field name
				name := tokens[i-1]
				names = append(names, name)
			}
		}
	}
	return names
}

func TestCompliancePrimitives(t *testing.T) {
	vectors := loadVectors[primitiveVector](t, "primitives.json")
	for _, v := range vectors {
		t.Run(v.Name, func(t *testing.T) {
			w := NewBitWriter()
			fields := extractFieldNames(v.Schema)
			for _, fn := range fields {
				val, ok := v.Value[fn]
				if !ok {
					continue
				}
				encodeField(t, w, v.Schema, fn, val)
			}
			got := w.Finish()
			want := hexToBytes(t, v.ExpectedBytes)
			if !bytesEqual(got, want) {
				t.Fatalf("got %X, want %X", got, want)
			}
		})
	}
}

func TestComplianceSubByte(t *testing.T) {
	vectors := loadVectors[subByteVector](t, "sub_byte.json")
	for _, v := range vectors {
		t.Run(v.Name, func(t *testing.T) {
			w := NewBitWriter()
			fields := extractFieldNames(v.Schema)
			for _, fn := range fields {
				val, ok := v.Value[fn]
				if !ok {
					continue
				}
				encodeField(t, w, v.Schema, fn, val)
			}
			got := w.Finish()
			want := hexToBytes(t, v.ExpectedBytes)
			if !bytesEqual(got, want) {
				t.Fatalf("got %X, want %X", got, want)
			}
		})
	}
}

func TestComplianceMessages(t *testing.T) {
	vectors := loadVectors[messageVector](t, "messages.json")
	for _, v := range vectors {
		t.Run(v.Name, func(t *testing.T) {
			w := NewBitWriter()
			fields := extractFieldNames(v.Schema)
			if len(fields) == 0 {
				// Empty message
				got := w.Finish()
				want := hexToBytes(t, v.ExpectedBytes)
				if !bytesEqual(got, want) {
					t.Fatalf("got %X, want %X", got, want)
				}
				return
			}
			for _, fn := range fields {
				val, ok := v.Value[fn]
				if !ok {
					continue
				}
				encodeField(t, w, v.Schema, fn, val)
			}
			got := w.Finish()
			want := hexToBytes(t, v.ExpectedBytes)
			if !bytesEqual(got, want) {
				t.Fatalf("got %X, want %X", got, want)
			}
		})
	}
}

func TestComplianceOptionals(t *testing.T) {
	vectors := loadVectors[primitiveVector](t, "optionals.json")
	for _, v := range vectors {
		t.Run(v.Name, func(t *testing.T) {
			w := NewBitWriter()
			val := v.Value["v"]
			if val == nil {
				w.WriteBool(false)
			} else {
				w.WriteBool(true)
				w.FlushToByteBoundary()
				w.WriteU32(uint32(toFloat64(val)))
			}
			got := w.Finish()
			want := hexToBytes(t, v.ExpectedBytes)
			if !bytesEqual(got, want) {
				t.Fatalf("got %X, want %X", got, want)
			}
		})
	}
}

func TestComplianceEnums(t *testing.T) {
	vectors := loadVectors[primitiveVector](t, "enums.json")
	variantMap := map[string]int{"Active": 0, "Inactive": 1}
	for _, v := range vectors {
		t.Run(v.Name, func(t *testing.T) {
			w := NewBitWriter()
			variant := v.Value["v"].(string)
			discriminant := variantMap[variant]
			w.WriteBits(uint64(discriminant), 1)
			got := w.Finish()
			want := hexToBytes(t, v.ExpectedBytes)
			if !bytesEqual(got, want) {
				t.Fatalf("got %X, want %X", got, want)
			}
		})
	}
}

func TestComplianceUnions(t *testing.T) {
	vectors := loadVectors[primitiveVector](t, "unions.json")
	for _, v := range vectors {
		t.Run(v.Name, func(t *testing.T) {
			w := NewBitWriter()
			unionVal := v.Value["v"].(map[string]interface{})
			variant := unionVal["variant"].(string)
			discriminant := 0
			if variant == "Rect" {
				discriminant = 1
			}

			pw := NewBitWriter()
			if variant == "Circle" {
				pw.WriteF32(float32(toFloat64(unionVal["radius"])))
			} else {
				pw.WriteF32(float32(toFloat64(unionVal["w"])))
				pw.WriteF32(float32(toFloat64(unionVal["h"])))
			}
			payload := pw.Finish()

			w.WriteLeb128(uint64(discriminant))
			w.WriteLeb128(uint64(len(payload)))
			w.WriteRawBytes(payload)
			got := w.Finish()
			want := hexToBytes(t, v.ExpectedBytes)
			if !bytesEqual(got, want) {
				t.Fatalf("got %X, want %X", got, want)
			}
		})
	}
}

func TestComplianceArraysMaps(t *testing.T) {
	vectors := loadVectors[primitiveVector](t, "arrays_maps.json")
	for _, v := range vectors {
		t.Run(v.Name, func(t *testing.T) {
			w := NewBitWriter()
			val := v.Value["v"]
			if arr, ok := val.([]interface{}); ok {
				w.WriteLeb128(uint64(len(arr)))
				for _, elem := range arr {
					w.WriteU32(uint32(toFloat64(elem)))
				}
			} else if m, ok := val.(map[string]interface{}); ok {
				entries := len(m)
				w.WriteLeb128(uint64(entries))
				for k, val := range m {
					w.WriteString(k)
					w.WriteU32(uint32(toFloat64(val)))
				}
			}
			got := w.Finish()
			want := hexToBytes(t, v.ExpectedBytes)
			if !bytesEqual(got, want) {
				t.Fatalf("got %X, want %X", got, want)
			}
		})
	}
}

func TestComplianceEvolution(t *testing.T) {
	data, err := os.ReadFile("../../compliance/vectors/evolution.json")
	if err != nil {
		t.Fatalf("failed to read evolution.json: %v", err)
	}
	var vectors []struct {
		Name        string                 `json:"name"`
		SchemaV1    string                 `json:"schema_v1"`
		SchemaV2    string                 `json:"schema_v2"`
		Type        string                 `json:"type"`
		ValueV1     map[string]interface{} `json:"value_v1"`
		ValueV2     map[string]interface{} `json:"value_v2"`
		EncodedV1   string                 `json:"encoded_v1"`
		EncodedV2   string                 `json:"encoded_v2"`
		DecodedAsV1 map[string]interface{} `json:"decoded_as_v1"`
		DecodedAsV2 map[string]interface{} `json:"decoded_as_v2"`
		Notes       string                 `json:"notes"`
	}
	if err := json.Unmarshal(data, &vectors); err != nil {
		t.Fatalf("failed to parse evolution.json: %v", err)
	}

	for _, v := range vectors {
		t.Run(v.Name, func(t *testing.T) {
			if v.Name == "v1_encode_v2_decode_appended_field" {
				w := NewBitWriter()
				w.WriteU32(uint32(toFloat64(v.ValueV1["x"])))
				got := w.Finish()
				want := hexToBytes(t, v.EncodedV1)
				if !bytesEqual(got, want) {
					t.Fatalf("v1 encode: got %X, want %X", got, want)
				}
			} else if v.Name == "v2_encode_v1_decode_trailing_ignored" {
				w := NewBitWriter()
				w.WriteU32(uint32(toFloat64(v.ValueV2["x"])))
				w.WriteU16(uint16(toFloat64(v.ValueV2["y"])))
				got := w.Finish()
				want := hexToBytes(t, v.EncodedV2)
				if !bytesEqual(got, want) {
					t.Fatalf("v2 encode: got %X, want %X", got, want)
				}
			}
		})
	}
}
